File: Hyerin Cheon/2022-07-29-Rotate-Image.js
// [1, 2, 3]
// [4, 5, 6]
// [7, 8, 9]


function rotate(matrix){
  for(let i = 0; i < matrix.length; i++){
    for(let j = i + 1; j < matrix[0].length; j++){
      temp = matrix[j][i];  //[1][0] 4
      matrix[j][i] = matrix[i][j]; // 4 --> 2
      matrix[i][j] = temp;  // 2 --> 4
    }
  }

  for(let row of matrix){
    row.reverse();
  }
}

File: MinjiJu/2022-summer/2022-07-29-rotate-image.cpp
// transpose => horizontal reflection
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // transpose
        for(int i=0; i<matrix.size(); i++){
            for(int j=0; j<i; j++){
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        // flip
        for(int i=0; i<matrix.size(); i++){
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

// vertical flip => transpose
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // flip
        reverse(matrix.begin(), matrix.end());
        // transpose
        for(int i=0; i<matrix.size(); ++i){
            for(int j=i+1; j<matrix[i].size(); ++j){
                swap(matrix[i][j], matrix[j][i]);
            }
        }
    }
};

// swaps
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int a=0, b=matrix.size()-1;
        while(a<b){
            for(int i=0; i<(b-a);++i){
                swap(matrix[a][a+i], matrix[a+i][b]);
                swap(matrix[a][a+i], matrix[b][b-i]);
                swap(matrix[a][a+i], matrix[b-i][a]);
            }
            ++a;--b;
        }
    }
};